lib/IDE/Refactoring.cpp

```diff
@@ -2066,6 +2073,107 @@ bool RefactoringActionSimplifyNumberLiteral::performChange() {
   return true;
 }
 
+static CallExpr *findTrailingClosureTarget(SourceManager &SM,
+                                           ResolvedCursorInfo CursorInfo) {
+  if (CursorInfo.Kind == CursorInfoKind::StmtStart)
+    return nullptr;
+
+  // Find inner most CallExpr
+  ContextFinder
+  Finder(*CursorInfo.SrcFile, CursorInfo.Loc,
+         [](ASTNode N) {
+           return N.isStmt(StmtKind::Brace) || N.isExpr(ExprKind::Call);
+         });
+  Finder.resolve();
+  if (Finder.getContexts().empty()
+      || !Finder.getContexts().back().is<Expr*>())
+    return nullptr;
+  CallExpr *CE = dyn_cast<CallExpr>(Finder.getContexts().back().get<Expr*>());
+
+  // The last arugment is a closure?
+  Expr *Args = CE->getArg();
+  if (!Args)
+    return nullptr;
+  Expr *LastArg;
+  if (auto *TSE = dyn_cast<TupleShuffleExpr>(Args))
+    Args = TSE->getSubExpr();
+  if (auto *PE = dyn_cast<ParenExpr>(Args)) {
+    LastArg = PE->getSubExpr();
+  } else {
+    auto *TE = cast<TupleExpr>(Args);
+    if (TE->getNumElements() == 0)
+      return nullptr;
+    LastArg = TE->getElements().back();
+  }
+
+  while (auto *ICE = dyn_cast<ImplicitConversionExpr>(LastArg))
```

Comment:
Can we check `isImplicit()` instead, which seems to be more generic.

Reply:
I think your original code makes sense, but can we add it as a member function to `ImplicitConversionExpr`?

Reply:
`isImplicit()` doesn't necessary mean it's `ImplicitConversionExpr`.
Probably, we should add a member function to `Expr`, say `getSingleSyntacticExpr()` which dig into `isImplicit()` expressions using `SourceEntitiyWalker`(or `ASTWalker`). (`Single` because implicit `TupleExpr` may contain multiple syntactic expressions. Returns `nullptr` in such cases).

Reply:
I think we can add `getSyntacticSubExpr()` to `ImplicitConversionExpr`. Adding such method to `Expr` seems to be an overkill because in a long term we will have libSyntax to get rid of all implicit nodes. Also, as you said, we have to return `nullptr` for `TupleExpr` which is a little ambiguous for we're not sure whether it means no explicit nodes or multiple of them. What do you think?

Reply:
I agree. Thanks!